infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/StateTransferManager.java

```diff
@@ -0,0 +1,316 @@
+                                .build();
+                    } else {
+                        // Get total number of addresses needed to transfer.
+                        long numAddressesToTransfer = unknownAddressesInRange.size();
+                        // Create a transferBatchRequest stream.
+                        Stream<TransferBatchRequest> batchStream = Lists
+                                .partition(unknownAddressesInRange, batchSize)
+                                .stream()
+                                .map(groupedAddresses -> new TransferBatchRequest(groupedAddresses, Optional.empty())
+                                );
+                        // Execute state transfer synchronously.
+                        newStatus = synchronousStateTransfer(batchProcessor, batchStream)
+                                .thenApply(result -> createStatusBasedOnTransferResult(result, numAddressesToTransfer))
+                                .join();
+
+                    }
+                    // If a segment contains some other status -> return.
+
+                    return TransferSegment
+                            .builder()
+                            .startAddress(segment.getStartAddress())
+                            .endAddress(segment.getEndAddress())
+                            .status(newStatus)
+                            .build();
+                }
+
+
+        ).collect(ImmutableList.toImmutableList());
+    }
+
+    /**
+     * Given a batch processor and a stream of batch requests, execute the state transfer.
+     *
+     * @param batchProcessor An instance of a batch processor.
+     * @param batchStream    A stream of batch requests.
+     * @return A completed future containing a result of total number of addresses transferred
+     * or an exception.
+     */
+    CompletableFuture<Result<Long, TransferSegmentFailure>> synchronousStateTransfer(
+            StateTransferBatchProcessor batchProcessor, Stream<TransferBatchRequest> batchStream) {
+        Result<Long, TransferSegmentFailure> accumulatedResult = Result.ok(0L);
+
+        Result<Long, TransferSegmentFailure> resultOfTransfer =
+                batchStream.reduce(accumulatedResult, (resultSoFar, nextBatch) -> {
+                    TransferBatchResponse transferBatchResponse =
+                            batchProcessor.transfer(nextBatch).join();
+                    if (transferBatchResponse.getStatus() == FailureStatus.FAILED) {
+                        return Result.error(new TransferSegmentFailure());
+                    } else {
+                        return resultSoFar
+                                .map(sumSoFar -> sumSoFar + transferBatchResponse
+                                        .getTransferBatchRequest()
+                                        .getAddresses()
+                                        .size());
+                    }
+                }, (oldSum, newSum) -> newSum);
```

Comment:
This implementation will continue to transfer the following batches even if one batch failed, and eventually you return a result with error. I think what you need is to stop and return a result with error immediately after a failure.

Reply:
Fixed